### PR TITLE
[12.0][ADD] account_document_reversal

### PR DESCRIPTION
**Pull request by kittiu**
By Odoo standard, when an account document is cancelled, its journal entry will be deleted completely.
This module enhance the process, instead of deletion, it will create new reversed journal entry.
This will help preserved the accounting history, which is strictly required by some country.

Following are documented provide this feature,

- Invoice (account.invoice)
- Payment (acccont.payment)
- Bank Statement (account.bank.statement.line)

### CODE REVIEW

**Comment by kittiu** [comment]
@jbeficent the test log show error on other module, can you check please.
https://travis-ci.org/OCA/account-financial-tools/jobs/533264019
```
************* Module account_move_fiscal_year
account_move_fiscal_year/readme/CONFIGURE.rst:1: [W7908(missing-newline-extrafiles), ]  Missing newline
account_move_fiscal_year/readme/DESCRIPTION.rst:1: [W7908(missing-newline-extrafiles), ]  Missing newline
account_move_fiscal_year/readme/INSTALL.rst:1: [W7908(missing-newline-extrafiles), ]  Missing newline
account_move_fiscal_year/readme/USAGE.rst:1: [W7908(missing-newline-extrafiles), ]  Missing newline
************* Module account_check_deposit.tests.test_check_deposit
account_check_deposit/tests/test_check_deposit.py:122: [W0622(redefined-builtin), 
```

Before I do this PR, I already get latest code from 12.0 and rebase.

**Comment by JordiBForgeFlow** [comment]
@kittiu you can ignore those messages from pylint. The overall status for the test_pytlint is success.

On the other hand you still have some outstanding Flake8 errors that are causing Travis to be red still:

image

**Comment by fclementic2c** [comment]
@kittiu Thats a tricky one.... but a real issue in Odoo.
What will happen if you confirm again a cancelled invoice or bank statement line ?
Do you plan to link 1 invoice, bk st. line to x account move lines ? 
Does it mean taht you disconnect the document sequence from teh account move sequence ?

**Comment by JordiBForgeFlow** [comment]
@fclementic2c when you cancel an invoice the original journal entry becomes unrelated to the invoice, and reversed. You are free to set the invoice back to draft, do any change required and validate again. This will create a brand new journal entry.

Same story goes with bank statement lines.

**Comment by kittiu** [comment]
> @kittiu you can ignore those messages from pylint. The overall status for the test_pytlint is success.
> 
> On the other hand you still have some outstanding Flake8 errors that are causing Travis to be red still:
> 
> image

Thank you Jordi. Fixed.

**Comment by kittiu** [comment]
@fclementic2c Yes as Jordi mentioned.

Actually, in the past that I implement this feature, and try to preserve the link and so "Set to Draft" can't be used. Invoice (or payment) must be re-created again, as the old one stay cancelled. It sound ideal, but it also result in many more use case we have to cover. After discuss more with Jordi, allow removing the link is not that of bad idea.

OK, the number of invoice can skip as result of re-confirm, but all the history journal entries still there. So far, I think it is balance between control and flexibility.

**Comment by fclementic2c** [comment]
@jbeficent Thanks for the explanation.
I see maybe one problem here: Today Odoo assume that the document (invoice, payment...) has the same number than the accounting entry in many places (which I have always been against).

ex: when you validate an invoice the invoice number is not stored on the journal entry.
What do you think ?

**Comment by JordiBForgeFlow** [comment]
@fclementic2c perhaps it would be interesting to store the invoice_id in the account.move, so that you always have the option to locate the journal entries for that invoice.

**Comment by kittiu** [comment]
@jbeficent Would you be interested to review this PR
https://github.com/OCA/account-financial-tools/pull/790

**Comment by kittiu** [comment]
This PR is now green :)

**Comment by fclementic2c** [comment]
Hello, I did a full test. 

Overall it works well but I have few feedbacks :

- About the cancellation wizard : 
* The wizard should not allow to reverse at the date before the invoice date
* I would not give the ability to post the reversal on a different journal.
* The button to confirm the cancellation is "Cancel" -> a bit confusing, you should change to "Confirm"
Once reversed, I loose the link the the original invoice and its reversal -> so I cannot check easily what odoo did for me. I would have appreciate a link to these entries from the invoice.

- Account_cancel is 'blocked' by the french certification module (l10n_fr_certification) because with this module you cannot activate the entries cancellation in journals -> 
error message: "According to the French law, you cannot modify a journal in order for its posted data to be updated or deleted. Unauthorized field: Allow Cancelling Entries."
So I will not be able to use this module for my french customers neither in any multi-company instance with at least 1 french company :\ 


Ergonomy:
When you post a manual entry (adviser/Journal entries): you see a button at the top of the screen : Reverse entry and a button Cancel entry -> a bit confusing to have a reversal and a cancel which does reversal too side-by-side.
But I could say this also for invoices.

And something I could not test... I am wondering if this will not break the Enterprise bank reconciliation) 

But Excellent job! :)

**Comment by kittiu** [comment]
@fclementic2c thank you so much for taking time testing, let's discuss more,

> About the cancellation wizard :
> The wizard should not allow to reverse at the date before the invoice date

OK, I agree, I will add blocking message.

> I would not give the ability to post the reversal on a different journal.

For this, I customers insisting they want all cancel journal entries in different journal. So, I think it might be a real need for some. So, I rather propose it to be configurable in the origin Journal like this. WDYT?
image

Note: further more, we can have another option "Disable Set to Draft", which will show blocking message when user click on Set to Draft. WDYT?

> The button to confirm the cancellation is "Cancel" -> a bit confusing, you should change to "Confirm"

OK, I agree. I feel the same.

> Once reversed, I loose the link the the original invoice and its reversal -> so I cannot check easily what odoo did for me. I would have appreciate a link to these entries from the invoice.

To maintain the simplicity, I think we rather leave it like this. But for the reference back, please consider this PR https://github.com/OCA/account-financial-tools/pull/790, which is to have document_id in account.move to link back to front document. We can then create smart button from the accounting document, i.e., (invoice, payment, etc) to show all related journal entries.

> Account_cancel is 'blocked' by the french certification module (l10n_fr_certification) because with this module you cannot activate the entries cancellation in journals ->
> error message: "According to the French law, you cannot modify a journal in order for its posted data to be updated or deleted. Unauthorized field: Allow Cancelling Entries."
> So I will not be able to use this module for my french customers neither in any multi-company instance with at least 1 french company :\

I will look more into this, to make this unrelated to account_cancel. I think we can.

@jbeficent please share opinion if you will.

**Comment by kittiu** [comment]
> Ergonomy

This once I don't really get your point. You can share some picture? Thanks!

**Comment by JordiBForgeFlow** [comment]
@fclementic2c AFAIKT if you are in france and use 'l10n_fr_certification' you can install the module 'account_cancel', and you will still get the error message.

Of course, it was back then stupid to install this module in France. Because you would get that error message.

This module does not use the 'button_cancel' button of account.move. You will still not be allowed to cancel moves with this module.

**Comment by JordiBForgeFlow** [comment]
@fclementic2c @kittiu I realized that when you specify in a journal 'allow cancelling entries', and you choose a method, then this method must be also applicable when you cancel a journal entry manually!

I have implemented the cancel wizard in account.move as well (in my backport to 11.0 https://github.com/OCA/account-financial-tools/pull/824)
https://github.com/OCA/account-financial-tools/pull/824/commits/45a258127e8bdf79d4ea8cf5667018cf240919a5

This makes much more sense to me, because it won't allow anyone to set to draft a posted journal entry. Instead the cancel wizard will be opened.

**Comment by JordiBForgeFlow** [comment]
@fclementic2c Then it all boils down to not allowing the cancel of journal entries to users, if you want to adhere to the french law, or better, use this module and make sure there are no loose ends.

**Comment by kittiu** [comment]
@jbeficent, noted on account.move will do the same here. What is your opinion on my propose to moving optional cancelling journal to journal setting instead of on the wizard?

**Comment by JordiBForgeFlow** [comment]
@kittiu I would set a flag in the journal 'Allow to set a canceling journal'. And then you display/hide in the wizard depending on that.

**Comment by kittiu** [comment]
@jbeficent I have seen your implementation. The Cancel button will show document cancel wizard, but at the end it do reversal as normal. I think it is confusing having 2 buttons, doing the same thing. I would rather show warning "This action is not allowed for cancel with reversal. Please use Reverse Entry instead", or hide the Cancel button. Better to do both.

**Comment by kittiu** [comment]
- [X] Add blocking and hiding for accout.move's button_cancel() for journal with cancel reversal
- [X] Change button name from "Cancel" to "Confirm"
- [X] Add checkbox "Use different journal for reversal"

@jbeficent @fclementic2c

**Comment by JordiBForgeFlow** [comment]
@kittiu In my opinion the check "Use different journal for reversal" should be in the journal, not in the widget.

image

image

So, if you flag this checkbox in the journal, then when the wizard is opened you will have the option.

**Comment by kittiu** [comment]
@jbeficent I misinterpret your word. Will fix that.

**Comment by kittiu** [comment]
Done. I also add option to set default reversal journal. Wizard will now show/hide according to this setup.

image

**Review by JordiBForgeFlow — APPROVED** [review]
:+1: Code and functional review. Works great!

**Comment by JordiBForgeFlow** [comment]
@kittiu In the future I'd appreciate if you can leave the commit history while reviewing. Otherwise it's hard to follow your changes. :)

**Comment by pedrobaeza** [comment]
@jbeficent no, this is better, as you don't need an extra step of squashing before merging. You can review changes clicking on the link on the words "forced push".

**Comment by kittiu** [comment]
Thank you for all comments and fixes. I will leave this as final for now. 
Note: ci/runbot is red, but this seem to be problem of runbot, perhaps?

**Comment by pedrobaeza** [comment]
Yes, runbot is having problems lately

**Comment by JordiBForgeFlow** [comment]
@kittiu Thanks. To me this is good to go.

@etobella if you want we can review together next week in Santiago.

**Comment by JordiBForgeFlow** [comment]
@fclementic2c Are you good with the changes?
